File: web/src/modules/scheduled-task/model/scheduledTask.ts
import { z } from 'zod';

const goSlice = <T extends z.ZodTypeAny>(item: T) => z.preprocess(
  (v) => (Array.isArray(v) ? v : []),
  z.array(item),
);

export const scheduledTaskSchema = z.object({
  id: z.string().min(1),
  name: z.string().min(1),
  workflowId: z.string().min(1),
  versionId: z.string().min(1),
  inputTemplate: z.record(z.string(), z.unknown()).default({}),
  cronExpr: z.string().min(1),
  enabled: z.boolean(),
  nextFireAt: z.string(),
  lastRunAt: z.string().optional(),
  lastRunStatus: z.string(),
  lastErrorMessage: z.string().optional(),
  createdBy: z.string(),
  createdAt: z.string(),
  updatedAt: z.string(),
  // 展示用引用实体名称（默认展示 name，原始 id 可 hover 悬浮）：引用实体
  // 已删除或未解析时后端缺省，前端回退展示原始 id。
  workflowName: z.string().optional(),
  versionNo: z.number().int().nonnegative().optional(),
  versionName: z.string().optional(),
}).strict();
export type ScheduledTask = z.infer<typeof scheduledTaskSchema>;

export const scheduledTaskPageSchema = z.object({
  tasks: goSlice(scheduledTaskSchema),
  total: z.number().int().nonnegative(),
  page: z.number().int().positive(),
  pageSize: z.number().int().positive(),
});
export type ScheduledTaskPage = z.infer<typeof scheduledTaskPageSchema>;

export interface CreateScheduledTaskInput {
  name: string;
  workflowId: string;
  versionId: string;
  inputTemplate: Record<string, unknown>;
  cronExpr: string;
}

export type UpdateScheduledTaskInput = CreateScheduledTaskInput;

File: api/http/dto/scheduled_task.go
package dto

import (
	"time"

	scheddomain "github.com/byteBuilderX/stratum/internal/scheduler/domain"
)

// CreateScheduledTaskRequest creates an enabled scheduled task bound to a
// workflow version. Cron validation and input-schema conformance are checked
// by the service so bad schedules surface as 400s before any row is written.
type CreateScheduledTaskRequest struct {
	Name          string         `json:"name" binding:"required"`
	WorkflowID    string         `json:"workflowId" binding:"required"`
	VersionID     string         `json:"versionId" binding:"required"`
	InputTemplate map[string]any `json:"inputTemplate"`
	CronExpr      string         `json:"cronExpr" binding:"required"`
}

// UpdateScheduledTaskRequest replaces every editable field; the schedule is
// restarted from the freshly computed next fire time.
type UpdateScheduledTaskRequest struct {
	Name          string         `json:"name" binding:"required"`
	WorkflowID    string         `json:"workflowId" binding:"required"`
	VersionID     string         `json:"versionId" binding:"required"`
	InputTemplate map[string]any `json:"inputTemplate"`
	CronExpr      string         `json:"cronExpr" binding:"required"`
}

// SetScheduledTaskEnabledRequest toggles the schedule. Re-enabling recomputes
// next_fire_at from now; disabling keeps the row but removes it from the due
// set.
//
// 注意：不能加 binding:"required"——bool 的零值 false 会被 required 拒绝。
type SetScheduledTaskEnabledRequest struct {
	Enabled bool `json:"enabled"`
}

// ScheduledTaskResponse is the admin/member surface of one scheduled task.
type ScheduledTaskResponse struct {
	ID               string         `json:"id"`
	Name             string         `json:"name"`
	WorkflowID       string         `json:"workflowId"`
	VersionID        string         `json:"versionId"`
	InputTemplate    map[string]any `json:"inputTemplate"`
	CronExpr         string         `json:"cronExpr"`
	Enabled          bool           `json:"enabled"`
	NextFireAt       time.Time      `json:"nextFireAt"`
	LastRunAt        *time.Time     `json:"lastRunAt,omitempty"`
	LastRunStatus    string         `json:"lastRunStatus"`
	LastErrorMessage string         `json:"lastErrorMessage,omitempty"`
	CreatedBy        string         `json:"createdBy"`
	CreatedAt        time.Time      `json:"createdAt"`
	UpdatedAt        time.Time      `json:"updatedAt"`
}

// ToScheduledTaskResponse maps a scheduled task to its HTTP surface.
func ToScheduledTaskResponse(t scheddomain.ScheduledTask) ScheduledTaskResponse {
	return ScheduledTaskResponse{
		ID:               t.ID,
		Name:             t.Name,
		WorkflowID:       t.WorkflowID,
		VersionID:        t.VersionID,
		InputTemplate:    t.InputTemplate,
		CronExpr:         t.CronExpr,
		Enabled:          t.Enabled,
		NextFireAt:       t.NextFireAt,
		LastRunAt:        t.LastRunAt,
		LastRunStatus:    t.LastRunStatus,
		LastErrorMessage: t.LastErrorMessage,
		CreatedBy:        t.CreatedBy,
		CreatedAt:        t.CreatedAt,
		UpdatedAt:        t.UpdatedAt,
	}
}

// ScheduledTaskPageResponse paginates the task list newest-first.
type ScheduledTaskPageResponse struct {
	Tasks    []ScheduledTaskResponse `json:"tasks"`
	Total    int                     `json:"total"`
	Page     int                     `json:"page"`
	PageSize int                     `json:"pageSize"`
}
